### PR TITLE
Move EnumChildWindows to Interop User32

### PR DESCRIPTION
**Pull request by gpetrou**
## Proposed changes

- Move EnumChildWindows to Interop User32.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2595)

### CODE REVIEW

File: src/Common/src/Interop/User32/Interop.EnumChildWindows.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class User32
+    {
+        public delegate bool EnumChildWindowsCallback(IntPtr hwnd, IntPtr lParam);
```
Comment (hughbe) [inline src/Common/src/Interop/User32/Interop.EnumChildWindows.cs]: I think this should be `WNDENUMPROC` from the docs

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs [hunk 2]
```diff
@@ -640,9 +640,10 @@ private unsafe static bool SendThemeChanged(IntPtr handle, IntPtr extraParameter
         private static bool SendThemeChangedRecursive(IntPtr handle, IntPtr lparam)
         {
             // First send to all children...
-            UnsafeNativeMethods.EnumChildWindows(new HandleRef(null, handle),
-                new NativeMethods.EnumChildrenCallback(Application.SendThemeChangedRecursive),
-                NativeMethods.NullHandleRef);
+            User32.EnumChildWindows(
+                handle,
+                new User32.EnumChildWindowsCallback(SendThemeChangedRecursive),
```
Comment (hughbe) [inline src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: @jeremykuhne do we need to do some sort of GC.KeepAlive thing here for delegates. I know it was like this betore

File: src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs [hunk 3]
```diff
@@ -1565,8 +1565,11 @@ private unsafe void UpdateUpDown()
             if (ShowUpDown)
             {
                 EnumChildren c = new EnumChildren();
-                NativeMethods.EnumChildrenCallback cb = new NativeMethods.EnumChildrenCallback(c.enumChildren);
-                UnsafeNativeMethods.EnumChildWindows(new HandleRef(this, Handle), cb, NativeMethods.NullHandleRef);
+                User32.EnumChildWindowsCallback cb = new User32.EnumChildWindowsCallback(c.enumChildren);
+                User32.EnumChildWindows(
+                    this,
+                    new User32.EnumChildWindowsCallback(c.enumChildren),
```
Comment (hughbe) [inline src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs]: Nit: isn’t this the same as `cb` Above?

File: src/Common/src/Interop/User32/Interop.EnumChildWindows.cs (hunk 1 above)
Comment (RussKie) [inline src/Common/src/Interop/User32/Interop.EnumChildWindows.cs]:
```suggestion
        public delegate BOOL EnumChildWindowsCallback(IntPtr hwnd, IntPtr lParam);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs (hunk 3 above)
Comment (weltkante) [reply src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs]: Yes, looks like a mistake (old code was using the variable and not allocating the delegate twice).

Also you'll probably want to verify whether you need GC.KeepAlive(cb) after the interop call to make sure the delegate doesn't get collected while the native code is running it, I'm not up to date with the interop rules here but the GC collecting delegates which are still running on the native side is something you have to be on the lookout for, its definitely an issue if the native side keeps the delegate longer than the function call. If its just using the delegate within the initial function call I don't know if keeping it alive is required.

Comment (weltkante) [reply src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs]: This [came up](https://github.com/dotnet/winforms/pull/2639#discussion_r363327044) in a different PR, keeping a delegate alive is [not required](https://docs.microsoft.com/en-us/dotnet/framework/interop/marshaling-a-delegate-as-a-callback-method) if it is not used past the end of the called function (i.e. its not stored/registered somewhere):

> When you use a delegate inside a call, the common language runtime protects the delegate from being garbage collected for the duration of that call.

**Comment by JeremyKuhne** [comment]
#2639 should be the one we take here as that also makes the callback static.

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs (hunk 2 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: I don't think so as the callbacks should be finished when `EnumChildWindows` returns.

**Comment by RussKie** [comment]
Thank you @gpetrou

**Pull request closed without merging**